Avoid account cache busting on page load

malithsen opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3692

## Changes proposed in this Pull Request:

When visiting the settings page, we clear the account cache, which forces an API request to Stripe to get a fresh copy of account data. This was added in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2833 to fix issues around statement descriptors. 

Fetching account data introduces a ~1s latency to the page load. This PR proposes the following changes to improve page load performance

1. Remove account refresh on page load (https://github.com/woocommerce/woocommerce-gateway-stripe/commit/e7a4662dd93ac1ce46edc2606034f225879342e5)
2. Listen to `account.updated` webhook event and clear the account cache if there's an update. This will force a new request on the next page refresh. (https://github.com/woocommerce/woocommerce-gateway-stripe/commit/e8cf0496d1656eb57a54b0e27c606c4bedfb1f5b)
3. Reconfigure webhook on plugin update (https://github.com/woocommerce/woocommerce-gateway-stripe/commit/e06cf600abe9a057a18cca77162e9390783be424)
4. Add a new menu item to manually refresh the account (https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3853/commits/46274f669be519d426c068f066d38aba55787b70)

3rd step is needed to enable `account.updated` event on existing webhooks. I am not certain if reconfiguring automatically on plugin update is the best way to do it. Another option is just to show a notice asking merchants to reconfigure manually. 
  
<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->
### Test page load speed improvement
1. On `develop` branch, go to settings page.
2. Open dev tools -> Network -> Doc 
3. Refresh the page and notice the timing for the `/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings` page
5. Switch to this branch and recheck the page load timing, it should be much faster on this branch.

### Test webhook
1. Make sure your local site is configured to receive webhooks from Stripe.
2. Go to Settings -> Account details -> Configure connection.
3. Click "Reconfigure webhooks" button.
4. Under "Customer bank statement" section in settings page, click on "Stripe account settings" link.
6. Update "Statement descriptor" or "Shortened descriptor" on Stripe.
7. Come back to the merchant site and refresh the page.
8. Updated info should be displayed under "Customer bank statement"

### Test webhook reconfiguration on plugin update
1. Manually bump the value in `wc_stripe_version` option in your db.
2. Go to WooCommerce -> Status -> Logs -> Stripe logs
3. Make sure there are following entries in the logs

### Test manual account refresh
1. Remove the webhook from the Stripe dashboard.
2. In Stripe, update "Statement descriptor" or "Shortened descriptor".
3. Come back to WC Stripe settings page and notice that the update statement descriptor is not yet reflected in the "Customer bank statement" section.
4. Click the "Refresh account details" menu item.
5. Notice that the "Customer bank statement" section gets updated with the latest data.

![CleanShot 2025-02-10 at 10 58 58](https://github.com/user-attachments/assets/1f088245-f127-40a9-b8b6-9c8cf7624767)


```
Successfully reconfigured webhooks for test mode after plugin update.
```

```
Successfully reconfigured webhooks for live mode after plugin update.
```
(Only available if you have live mode configured)
<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

malithsen committed:
Prevent account cache clear on settings page load

Clearing the cache forces fetching account data from Stripe. This
introduces ~1s delay in page load. So we remove it here to introduce
another mechanism fetch account data with a lower page load impact.

malithsen committed:
Clear account cache on webhook event

This forces a refetch of updated account data on the next page refresh

malithsen committed:
Reconfigure webhooks on plugin update

malithsen committed:
Merge branch 'develop' into tweak/avoid-account-cache-busting-on-page-load

malithsen committed:
Add changelog entries

malithsen commented on woocommerce-gateway-stripe.php:
```diff
@@ -366,6 +366,10 @@ public function install() {
 
+					// Add webhook reconfiguration
+					$account = self::get_instance()->account;
+					$account->reconfigure_webhooks_on_update();
```
I suppose we could be more restrictive here. Only reconfigure if there's an existing webhook and it listens to a different set of `enabled_events`.

diegocurbelo replied:
I agree... it's a bit more work, but it makes sense to only update the existing webhook if needed; we won't be adding events frequently.

malithsen replied:
@diegocurbelo Updated in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3853/commits/7a42ac72ab994be00867d4e0872bfe43f4afc93b :)

diegocurbelo committed:
Merge branch 'develop' into tweak/avoid-account-cache-busting-on-page-load

malithsen committed:
Only reconfigure webhooks when events differ.

- Add constant `WEBHOOK_EVENTS` to centralize webhook event list
- Enhance `reconfigure_webhooks_on_update()` to:
  * Check for existing webhooks
  * Compare current and desired webhook events
  * Only update webhooks when events differ

malithsen committed:
Add unit tests

malithsen committed:
Merge branch 'develop' into tweak/avoid-account-cache-busting-on-page-load

malithsen committed:
Change function name

diegocurbelo committed:
Merge branch 'develop' into tweak/avoid-account-cache-busting-on-page-load

diegocurbelo reviewed (approved):
Thank you for working on this, the 2 secs saved (with both changes) is awesome 😮 

The code looks good, and it tests well 🚢 

---
Just an idea...  but we might want to add an option to manually refresh the account in the account details options; that might help HEs (and us) to quickly detect/solve cache problems in the future.

Happy to create an issue for this and handle it in a different PR if you prefer.

![Screenshot_2025-02-07_at_17_05_12](https://github.com/user-attachments/assets/ea977881-0e95-4afc-9a56-b08a95c125f1)

malithsen committed:
Merge branch 'develop' into tweak/avoid-account-cache-busting-on-page-load

malithsen commented:
> we might want to add an option to manually refresh the account in the account details options; that might help HEs (and us) to quickly detect/solve cache problems in the future.

@diegocurbelo That's a great suggestion. ~I'll open a separate PR shortly.~ EDIT: nvm, looks like it's straightforward to implement. I'll do it in this PR.

BTW should we update the changelog entries with a new version? I see that the pre-release was created last Friday, which is what GS tested today. So I guess this change will not make it to the release this week, and the changelog should be updated accordingly?

malithsen committed:
Add a menu item to refresh the account details

malithsen committed:
Fix tests

Also remove a redundant test

malithsen commented:
@diegocurbelo I added the option to manually refresh account data in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3853/commits/46274f669be519d426c068f066d38aba55787b70

Also updated the test instructions to include "Test manual account refresh" section.

diegocurbelo reviewed (changes requested):
Thank you for adding the manual refresh option... it works perfectly 👏🏼 

One last detail; there is no feedback after clicking the "Refresh account details". How about adding a notice like the one we show when saving the settings: "Account details reloaded" or similar? 

![Screenshot 2025-02-10 at 16 49 53](https://github.com/user-attachments/assets/921c0a0f-2b17-44d4-8fd5-c280ffa19e70)

Or maybe triggering a page reload might be easier (it's not going to be a frequently used flow anyway)

malithsen committed:
Show notice when refresh is complete.

diegocurbelo reviewed (approved):
Thank for the changes @malithsen 👏🏼 

The code changes look great, and everything works as expected!! 🚢 

![Screenshot 2025-02-10 at 17 36 00](https://github.com/user-attachments/assets/a6f394cb-7ae5-433b-ba30-a6d915a8ab54)

malithsen commented:
@diegocurbelo Thanks for the review and improvement suggestions. 

Just double checking, should we update the changelog entries in this PR to point to 9.3.0? I see that the pre-release was created last Friday, so the build that GS tested didn't include this PR so I am guessing this PR will not be included in the release this week.

diegocurbelo commented:
> @diegocurbelo Thanks for the review and improvement suggestions.
> 
> Just double checking, should we update the changelog entries in this PR to point to 9.3.0? I see that the pre-release was created last Friday, so the build that GS tested didn't include this PR so I am guessing this PR will not be included in the release this week.

We usually don't update them until after the release (makes it easier to cherry-pick this PR into the release if needed).
Also, after this Thursday's release, we are going to merge @wjrosa's PR https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3548 to automate the changelog generation, so we'll need to update all 9.3.0 entries anyway 🤷🏼 ... keeping 9.2.0 makes sense.

malithsen commented:
Thanks. Good to know. Merging as-is then.
